### DISCUSSION TITLE
What is the proper way to extend a schema that adds additional properties to `provider:`?

### DISCUSSION THREAD

**Discussion by neverendingqs** [serverless/serverless#10307] [Q&A]
I currently maintain a plugin ([serverless-stack-policy-by-resource-type](https://github.com/neverendingqs/serverless-stack-policy-by-resource-type)) for the Serverless Framework that takes in additional properties for `provider.stackPolicy[x].ResourceType` and `provider.stackPolicy[x].ExcludeResource` (where `x` is for any element in that array). It causes the following warning when used:

```
Serverless:   at 'provider.stackPolicy[1]': unrecognized property 'ResourceType'
Serverless:   at 'provider.stackPolicy[1]': unrecognized property 'ExcludeResource'
Serverless:   at 'provider.stackPolicy[1]': unsupported configuration format
```

Should I use [defineProvider() helper](https://www.serverless.com/framework/docs/guides/plugins/custom-configuration#new-provider-via-defineprovider) to extend the schema? The documentation says it's for a new provider, so I'm not sure if it will simply merge if I provide an existing provider. E.g.

```js
serverless.configSchemaHandler.defineProvider('aws', { ... });
```

Side question: I couldn't find the code that implements schema validation when [searching for `defineProvider`](https://github.com/search?q=org%3Aserverless+defineProvider&type=code). Is the schema validation code in another GitHub Organization?

EDIT: also affects [serverless-plugin-lambda-account-access](https://github.com/rschick/serverless-plugin-lambda-account-access/pull/19).

**Comment by mnapoli** [comment]
Pinging @medikoo or @pgrzesik for help, I'm not sure if that will indeed work.

**Comment by medikoo** [comment]
Apparently, we neither documented that nor have a dedicated method. It's surprising we're discovering it now.

Anyway it can be achieved via public API as:

```javascript
serverless.configSchemaHandler.addPropertiesToSchema(
  serverless.configSchemaHandler.schema.properties.provider,
  { ... } // Schema extension
)
```

**Reply by neverendingqs** [reply]
Thanks! Will give it a try and report back.

**Reply by neverendingqs** [reply]
Gave it a try and got the following error:

```
 Type Error ----------------------------------------------

  TypeError: serverless.configSchemaHandler.addPropertiesToSchema is not a function
      at new StackPolicyByResourceType (/home/neverendingqs/code/neverendingqs/serverless-stack-policy-by-resource-type/src/index.js:5:36)
      at PluginManager.addPlugin (/home/neverendingqs/code/sls-sandbox/node_modules/serverless/lib/classes/PluginManager.js:114:28)
      at /home/neverendingqs/code/sls-sandbox/node_modules/serverless/lib/classes/PluginManager.js:163:33
      at Array.forEach (<anonymous>)
      at PluginManager.loadAllPlugins (/home/neverendingqs/code/sls-sandbox/node_modules/serverless/lib/classes/PluginManager.js:163:8)
      at Serverless.init (/home/neverendingqs/code/sls-sandbox/node_modules/serverless/lib/Serverless.js:213:30)
      at async /home/neverendingqs/code/sls-sandbox/node_modules/serverless/scripts/serverless.js:585:7
```

It looks like `addPropertiesToSchema()` is defined as a separate function and not in `ConfigSchemaHandler`:https://github.com/serverless/serverless/blob/e4431f26c12682a26bcea7233c61645088434f3b/lib/classes/ConfigSchemaHandler/index.js#L404-L422

(In contrast, `defineFunctionEvent()` is defined in `ConfigSchemaHandler`: https://github.com/serverless/serverless/blob/e4431f26c12682a26bcea7233c61645088434f3b/lib/classes/ConfigSchemaHandler/index.js#L274-L302).

If that's the case, would there be interest in changing `addPropertiesToSchema()` to be a class method of `ConfigSchemaHandler` @medikoo? If so I can attempt a PR.

**Reply by medikoo** [reply]
> , would there be interest in changing addPropertiesToSchema() to be a class method of ConfigSchemaHandler @medikoo? If so I can attempt a PR.

Yes, @neverendingqs, if you can prepare a PR, we'll be grateful. Thanks!

**Reply by neverendingqs** [reply]
I gave this a try:

* `serverless/serverless` change: https://github.com/serverless/serverless/pull/10341
* Plugin change: https://github.com/neverendingqs/serverless-stack-policy-by-resource-type/pull/1

I ran into issues because:

* `stackPolicy` already exists, so `addPropertiesToSchema()` will throw an error
* `stackPolicy` is a reference to another schema (`{ '$ref': '#/definitions/awsIamPolicyStatements' }`), so even if we changed `addPropertiesToSchema()` to allow appends, we need to do some sort of dereferencing first, and I don't think any `serverless/serverless` code currently exists to do something like this

I think there is some value in proceeding with #10341, but I don't think it'll solve what I'm looking to do above.

Thoughts @medikoo?

**Comment by medikoo** [comment]
@neverendingqs I see the issue. In this case I think the proper way would be if we keep `stackPolicy` setting only for IAM compliant input (as it's configured in a core), and introduce custom handling via other property e.g. plugin could add support for something as:

```yaml
provider:
  iam:
    stackPolicyByType:
      - ResourceType: ..
        Effect: ..

**Reply by neverendingqs** [reply]
Thanks. I'll need to think about it, as this will cause a breaking change to the plugin.

The plugin uses the new properties to convert it into a valid `stackPolicy` before CloudFormation deploy occurs, so it doesn't cause problems today.

**Reply by medikoo** [reply]
Technically you can also try to alter schema directly (it's accessible at `serverless.configSchemaHandler.schema`). Those helper methods were created simply for convenience as otherwise it'll be challenging for plugin authors to come up with extensions. 

Anyway, concerning this specific case, I think safer pattern would be if plugin do not encourage to decorate parts that are passed to CloudFormation template directly. It'll be better to maintain the separation, between plugin config and what goes into CF template.

e.g. configuration as follows could work well:

```yaml
provider:
  iam:
    stackPolicyByType:
      - type: AWS::S3::Bucket
      - policy:
          Effect: Allow
          Principal: '*'
          Action: 'Update:*'

```
